File: pastperfect/sources/rijksmuseum.py
"""Rijksmuseum — Linked Art data service.

https://data.rijksmuseum.nl/docs/search — no key required. The collection is
published as Linked Art, so a single object takes three hops: the object record
names a VisualItem, which carries the rights statement and names a DigitalObject,
which finally carries the IIIF image URL. Responses are cached on disk, so the
cost is paid once per object.
"""

from __future__ import annotations

import re
from collections.abc import Iterator

from .base import BlockedError, RawObject, clean, fetch_json

MUSEUM = "rijksmuseum"
SEARCH = "https://data.rijksmuseum.nl/search/collection"

AAT_EN = "http://vocab.getty.edu/aat/300388277"
AAT_TITLE = "http://vocab.getty.edu/aat/300404670"
AAT_CREDIT = "http://vocab.getty.edu/aat/300026687"
AAT_DESCRIPTION = "http://vocab.getty.edu/aat/300435416"
AAT_WEBPAGE = "http://vocab.getty.edu/aat/300264578"

#: A spread of object types, so the game is not all prints. Rijksmuseum search
#: accepts English or Dutch terms for these.
TYPES = [
    "painting", "drawing", "photograph", "sculpture", "furniture", "print",
    "glass", "textile", "jewellery", "silver", "ceramic", "costume",
]


def _centuries(windows: list[tuple[int, int]]) -> list[str]:
    """Rijksmuseum search filters dates by wildcard, so windows become centuries."""
    out: list[str] = []
    for begin, end in windows:
        for year in range(max(begin, 1000), min(end, 2100), 100):
            token = f"{year // 100:02d}??"
            if token not in out:
                out.append(token)
    return out


def harvest(windows: list[tuple[int, int]], per_window: int, seed: int = 7) -> Iterator[RawObject]:
    """Round-robin across century x object-type, stopping at the overall target.

    Each object costs three requests, so the walk is breadth-first: one pass
    takes a couple of objects from every combination before any combination gets
    a second look. That keeps a small quota spread across the whole collection.
    """
    total = max(1, per_window * len(windows))
    combos = [(century, kind) for century in _centuries(windows) for kind in TYPES]
    if not combos:
        return
    listings: dict[tuple[str, str], list[str]] = {}
    yielded = 0
    per_pass = 2
    offset = 0
    while yielded < total and offset < 40:
        progressed = False
        for combo in combos:
            if yielded >= total:
                return
            if combo not in listings:
                century, kind = combo
                url = f"{SEARCH}?type={kind}&imageAvailable=True&creationDate={century}"
                try:
                    payload = fetch_json(url, museum=MUSEUM)
                except BlockedError:
                    raise
                except Exception:
                    listings[combo] = []
                    continue
                listings[combo] = [
                    item.get("id") for item in _dicts(payload or {}, "orderedItems")
                    if item.get("id")
                ]
            ids = listings[combo][offset : offset + per_pass]
            for object_id in ids:
                progressed = True
                record = _load_object(object_id)
                if record is None:
                    continue
                yielded += 1
                yield record
                if yielded >= total:
                    return
        if not progressed:
            return
        offset += per_pass


# --- Linked Art helpers ---------------------------------------------------


def _as_list(value) -> list:
    """Linked Art serialises a single value as an object, not a one-item array.

    Every traversal below goes through this, so a record with one title behaves
    exactly like a record with three.
    """
    if value is None:
        return []
    if isinstance(value, list):
        return value
    return [value]


def _dicts(node, key: str) -> list[dict]:
    if not isinstance(node, dict):
        return []
    return [item for item in _as_list(node.get(key)) if isinstance(item, dict)]


def _notation(node, prefer_en: bool = True) -> str | None:
    if not isinstance(node, dict):
        return clean(node) if isinstance(node, str) else None
    values = _as_list(node.get("notation"))
    if not values:
        return clean(node.get("_label"))
    english = [v for v in values if isinstance(v, dict) and v.get("@language") == "en"]
    pool = english if (prefer_en and english) else values
    for value in pool:
        if isinstance(value, dict) and clean(value.get("@value")):
            return clean(value.get("@value"))
        if isinstance(value, str) and clean(value):
            return clean(value)
    return None


def _is_english(node: dict) -> bool:
    return any(lang.get("id") == AAT_EN for lang in _dicts(node, "language"))


def _classified(node: dict, aat: str) -> bool:
    return any(c.get("id") == aat for c in _dicts(node, "classified_as"))


def _title(obj: dict) -> str | None:
    names = [n for n in _dicts(obj, "identified_by") if n.get("type") == "Name"]
    for wanted_english in (True, False):
        for name in names:
            if not _classified(name, AAT_TITLE):
                continue
            if wanted_english and not _is_english(name):
                continue
            if clean(name.get("content")):
                return clean(name.get("content"))
    for name in names:
        if clean(name.get("content")):
            return clean(name.get("content"))
    return None


def _referred(obj, aat: str) -> str | None:
    notes = [n for n in _dicts(obj, "referred_to_by") if _classified(n, aat)]
    for wanted_english in (True, False):
        for note in notes:
            if wanted_english and not _is_english(note):
                continue
            if clean(note.get("content")):
                return clean(note.get("content"))
    return None


def _year(iso: str | None) -> int | None:
    if not iso:
        return None
    text = str(iso)
    negative = text.startswith("-")
    digits = text.lstrip("-").split("-")[0]
    if not digits.isdigit():
        return None
    return -int(digits) if negative else int(digits)


def _productions(obj: dict) -> list[dict]:
    produced = obj.get("produced_by")
    produced = produced if isinstance(produced, dict) else {}
    return [produced, *_dicts(produced, "part")] if produced else []


#: Names that mean "we do not know", which the interface should leave blank
#: rather than print as though somebody were called this.
_UNNAMED = {"anonymous", "anoniem", "onbekend", "unknown", "unidentified"}


def _creator(obj: dict) -> str | None:
    for production in _productions(obj):
        for actor in _dicts(production, "carried_out_by"):
            name = _notation(actor)
            if name:
                return name
    produced = obj.get("produced_by")
    return _referred(produced, AAT_DESCRIPTION) if isinstance(produced, dict) else None


def _page_url(obj: dict) -> str | None:
    for note in _dicts(obj, "subject_of"):
        for carrier in _dicts(note, "digitally_carried_by"):
            if carrier.get("format") != "text/html":
                continue
            for point in _dicts(carrier, "access_point"):
                url = clean(point.get("id"))
                if url:
                    return url
    return None


def _artist_name(obj: dict) -> str | None:
    name = clean(re.sub(
        r"\s*\((?:mentioned on object|possibly|attributed to)[^)]*\)", "", _creator(obj) or ""
    ))
    if not name or name.lower() in _UNNAMED:
        return None
    return name


def _load_object(object_id: str | None) -> RawObject | None:
    if not object_id:
        return None
    try:
        obj = fetch_json(object_id, museum=MUSEUM)
    except BlockedError:
        raise
    except Exception:
        return None
    if not obj:
        return None

    title = _title(obj)
    if not title:
        return None

    visual_ids = [v.get("id") for v in _dicts(obj, "shows") if v.get("id")]
    if not visual_ids:
        return None
    try:
        visual = fetch_json(visual_ids[0], museum=MUSEUM)
    except BlockedError:
        raise
    except Exception:
        return None
    if not visual:
        return None

    licence_raw = None
    for right in _dicts(visual, "subject_to"):
        for kind in _dicts(right, "classified_as"):
            if "creativecommons.org" in str(kind.get("id", "")):
                licence_raw = kind.get("id")
                break
    digital_ids = [d.get("id") for d in _dicts(visual, "digitally_shown_by") if d.get("id")]
    if not digital_ids:
        return None
    try:
        digital = fetch_json(digital_ids[0], museum=MUSEUM)
    except BlockedError:
        raise
    except Exception:
        return None
    image_url = None
    for point in _dicts(digital or {}, "access_point"):
        if clean(point.get("id")):
            image_url = clean(point.get("id"))
            break
    if not image_url:
        return None
    # The IIIF endpoint serves any width; ask for something a browser can use.
    image_url = image_url.replace("/full/max/", "/full/1100,/")

    produced = obj.get("produced_by")
    produced = produced if isinstance(produced, dict) else {}
    timespan = produced.get("timespan")
    timespan = timespan if isinstance(timespan, dict) else {}
    date_display = ""
    for name in _dicts(timespan, "identified_by"):
        if clean(name.get("content")):
            date_display = clean(name.get("content")) or ""
            break

    object_type = None
    for kind in _dicts(obj, "classified_as"):
        object_type = _notation(kind) or object_type
    materials = [m for m in (_notation(x) for x in _dicts(obj, "made_of")) if m]
    place = None
    for production in _productions(obj):
        for spot in _dicts(production, "took_place_at"):
            place = place or _notation(spot, prefer_en=False)

    source_id = str(object_id).rstrip("/").rsplit("/", 1)[-1]
    page = _page_url(obj) or f"https://id.rijksmuseum.nl/{source_id}"
    page = page.replace("rijksmuseum.nl/nl/", "rijksmuseum.nl/en/")
    return RawObject(
        museum=MUSEUM,
        source_id=source_id,
        title=title,
        object_url=page,
        image_url=image_url,
        licence_raw=licence_raw,
        rights_basis="rights statement on the object's VisualItem in the Rijksmuseum Linked Art data",
        date_display=date_display,
        year_start=_year(timespan.get("begin_of_the_begin")),
        year_end=_year(timespan.get("end_of_the_end")),
        artist=_artist_name(obj),
        artist_note=None,
        medium=", ".join(materials[:3]) or None,
        classification=object_type,
        culture=place,
        department=None,
        credit_line=_referred(obj, AAT_CREDIT) or "Rijksmuseum, Amsterdam",
        extra={},
    )

File: src/sources/rijksmuseum.ts
/**
 * Rijksmuseum — Linked Art data service.
 *
 * https://data.rijksmuseum.nl/docs/search — no key required. The collection is
 * published as Linked Art, so a single object takes three hops: the object
 * record names a VisualItem, which carries the rights statement and names a
 * DigitalObject, which finally carries the IIIF image URL. Responses are cached
 * on disk, so the cost is paid once per object.
 */

import { BlockedError, clean, fetchJson, raw, type RawObject } from "./base.ts";

export const MUSEUM = "rijksmuseum";
const SEARCH = "https://data.rijksmuseum.nl/search/collection";

const AAT_EN = "http://vocab.getty.edu/aat/300388277";
const AAT_TITLE = "http://vocab.getty.edu/aat/300404670";
const AAT_CREDIT = "http://vocab.getty.edu/aat/300026687";
const AAT_DESCRIPTION = "http://vocab.getty.edu/aat/300435416";

/**
 * A spread of object types, so the game is not all prints. Rijksmuseum search
 * accepts English or Dutch terms for these.
 */
const TYPES = [
  "painting", "drawing", "photograph", "sculpture", "furniture", "print",
  "glass", "textile", "jewellery", "silver", "ceramic", "costume",
];

/** Names that mean "we do not know", which the interface should leave blank. */
const UNNAMED = new Set(["anonymous", "anoniem", "onbekend", "unknown", "unidentified"]);

type Node = Record<string, unknown>;

/**
 * Linked Art serialises a single value as an object, not a one-item array.
 * Every traversal below goes through this, so a record with one title behaves
 * exactly like a record with three.
 */
function asList(value: unknown): unknown[] {
  if (value === null || value === undefined) return [];
  return Array.isArray(value) ? value : [value];
}

function dicts(node: unknown, key: string): Node[] {
  if (typeof node !== "object" || node === null) return [];
  return asList((node as Node)[key]).filter(
    (item): item is Node => typeof item === "object" && item !== null,
  );
}

function notation(node: unknown, preferEn = true): string | null {
  if (typeof node === "string") return clean(node);
  if (typeof node !== "object" || node === null) return null;
  const values = asList((node as Node)["notation"]);
  if (values.length === 0) return clean((node as Node)["_label"]);
  const english = values.filter(
    (v) => typeof v === "object" && v !== null && (v as Node)["@language"] === "en",
  );
  const pool = preferEn && english.length > 0 ? english : values;
  for (const value of pool) {
    if (typeof value === "object" && value !== null) {
      const text = clean((value as Node)["@value"]);
      if (text) return text;
    } else if (typeof value === "string" && clean(value)) {
      return clean(value);
    }
  }
  return null;
}

const isEnglish = (node: Node): boolean =>
  dicts(node, "language").some((lang) => lang["id"] === AAT_EN);

const classified = (node: Node, aat: string): boolean =>
  dicts(node, "classified_as").some((c) => c["id"] === aat);

function title(obj: Node): string | null {
  const names = dicts(obj, "identified_by").filter((n) => n["type"] === "Name");
  for (const wantEnglish of [true, false]) {
    for (const name of names) {
      if (!classified(name, AAT_TITLE)) continue;
      if (wantEnglish && !isEnglish(name)) continue;
      const text = clean(name["content"]);
      if (text) return text;
    }
  }
  for (const name of names) {
    const text = clean(name["content"]);
    if (text) return text;
  }
  return null;
}

function referred(obj: unknown, aat: string): string | null {
  const notes = dicts(obj, "referred_to_by").filter((n) => classified(n, aat));
  for (const wantEnglish of [true, false]) {
    for (const note of notes) {
      if (wantEnglish && !isEnglish(note)) continue;
      const text = clean(note["content"]);
      if (text) return text;
    }
  }
  return null;
}

function yearOf(iso: unknown): number | null {
  if (!iso) return null;
  const text = String(iso);
  const negative = text.startsWith("-");
  const digits = text.replace(/^-/, "").split("-")[0]!;
  if (!/^\d+$/.test(digits)) return null;
  return negative ? -Number(digits) : Number(digits);
}

function productions(obj: Node): Node[] {
  const produced = obj["produced_by"];
  if (typeof produced !== "object" || produced === null) return [];
  return [produced as Node, ...dicts(produced, "part")];
}

function creator(obj: Node): string | null {
  for (const production of productions(obj)) {
    for (const actor of dicts(production, "carried_out_by")) {
      const name = notation(actor);
      if (name) return name;
    }
  }
  return referred(obj["produced_by"], AAT_DESCRIPTION);
}

function artistName(obj: Node): string | null {
  const name = clean(
    (creator(obj) ?? "").replace(/\s*\((?:mentioned on object|possibly|attributed to)[^)]*\)/g, ""),
  );
  if (!name || UNNAMED.has(name.toLowerCase())) return null;
  return name;
}

/**
 * Repair a Rijksmuseum object page URL.
 *
 * The Linked Art data still publishes the museum's *previous* public URL --
 * `/en/collectie/object/<object number>--<hash>` -- and every one of those now
 * returns 404. The live scheme is `/en/collection/<object number>`, which
 * redirects to whatever slug the museum currently uses, so it keeps working
 * when they rename a thing.
 *
 * The object number is the only durable part, and the dead URL still contains
 * it. Anything this cannot recognise returns "" rather than a guess: an empty
 * `object_url` renders as no link at all, which is honest, and a link to a 404
 * is not.
 */
export function objectPageUrl(published: string | null | undefined): string {
  const url = (published ?? "").trim();
  if (!url) return "";
  // Already the live scheme.
  const live = /^https?:\/\/(?:www\.)?rijksmuseum\.nl\/en\/collection\/[^/?#]+$/;
  if (live.test(url)) return url;
  const dead = /\/(?:collectie|collection)\/object\/([^/?#]+?)(?:--[0-9a-f]{8,})?$/;
  const number = dead.exec(url)?.[1];
  if (!number) return "";
  return `https://www.rijksmuseum.nl/en/collection/${number}`;
}

function pageUrl(obj: Node): string | null {
  for (const note of dicts(obj, "subject_of")) {
    for (const carrier of dicts(note, "digitally_carried_by")) {
      if (carrier["format"] !== "text/html") continue;
      for (const point of dicts(carrier, "access_point")) {
        const url = clean(point["id"]);
        if (url) return url;
      }
    }
  }
  return null;
}

/** Rijksmuseum search filters dates by wildcard, so windows become centuries. */
function centuries(windows: ReadonlyArray<readonly [number, number]>): string[] {
  const out: string[] = [];
  for (const [begin, end] of windows) {
    for (let year = Math.max(begin, 1000); year < Math.min(end, 2100); year += 100) {
      const token = String(Math.floor(year / 100)).padStart(2, "0") + "??";
      if (!out.includes(token)) out.push(token);
    }
  }
  return out;
}

async function loadObject(objectId: string | undefined): Promise<RawObject | null> {
  if (!objectId) return null;
  const obj = (await fetchJson(objectId, { museum: MUSEUM })) as Node | null;
  if (!obj) return null;

  const name = title(obj);
  if (!name) return null;

  const visualIds = dicts(obj, "shows").map((v) => v["id"]).filter(Boolean) as string[];
  if (visualIds.length === 0) return null;
  const visual = (await fetchJson(visualIds[0]!, { museum: MUSEUM })) as Node | null;
  if (!visual) return null;

  let licenceRaw: string | null = null;
  for (const right of dicts(visual, "subject_to")) {
    for (const kind of dicts(right, "classified_as")) {
      if (String(kind["id"] ?? "").includes("creativecommons.org")) {
        licenceRaw = String(kind["id"]);
        break;
      }
    }
  }

  const digitalIds = dicts(visual, "digitally_shown_by").map((d) => d["id"]).filter(Boolean) as string[];
  if (digitalIds.length === 0) return null;
  const digital = (await fetchJson(digitalIds[0]!, { museum: MUSEUM })) as Node | null;
  let imageUrl: string | null = null;
  for (const point of dicts(digital ?? {}, "access_point")) {
    const url = clean(point["id"]);
    if (url) {
      imageUrl = url;
      break;
    }
  }
  if (!imageUrl) return null;
  // The IIIF endpoint serves any width; ask for something a browser can use.
  imageUrl = imageUrl.replace("/full/max/", "/full/1100,/");

  const produced = typeof obj["produced_by"] === "object" && obj["produced_by"] !== null
    ? (obj["produced_by"] as Node) : {};
  const timespan = typeof produced["timespan"] === "object" && produced["timespan"] !== null
    ? (produced["timespan"] as Node) : {};
  let dateDisplay = "";
  for (const label of dicts(timespan, "identified_by")) {
    const text = clean(label["content"]);
    if (text) {
      dateDisplay = text;
      break;
    }
  }

  let objectType: string | null = null;
  for (const kind of dicts(obj, "classified_as")) objectType = notation(kind) ?? objectType;
  const materials = dicts(obj, "made_of").map((m) => notation(m)).filter((m): m is string => Boolean(m));
  let place: string | null = null;
  for (const production of productions(obj)) {
    for (const spot of dicts(production, "took_place_at")) place ??= notation(spot, false);
  }

  const sourceId = String(objectId).replace(/\/+$/, "").split("/").pop()!;
  const page = objectPageUrl(
    (pageUrl(obj) ?? "").replace("rijksmuseum.nl/nl/", "rijksmuseum.nl/en/"),
  );

  return raw({
    museum: MUSEUM,
    source_id: sourceId,
    title: name,
    object_url: page,
    image_url: imageUrl,
    licence_raw: licenceRaw,
    rights_basis: "rights statement on the object's VisualItem in the Rijksmuseum Linked Art data",
    date_display: dateDisplay,
    year_start: yearOf(timespan["begin_of_the_begin"]),
    year_end: yearOf(timespan["end_of_the_end"]),
    artist: artistName(obj),
    medium: materials.slice(0, 3).join(", ") || null,
    classification: objectType,
    culture: place,
    credit_line: referred(obj, AAT_CREDIT) ?? "Rijksmuseum, Amsterdam",
  });
}

/**
 * Round-robin across century x object-type, stopping at the overall target.
 *
 * Each object costs three requests, so the walk is breadth-first: one pass
 * takes a couple of objects from every combination before any combination gets
 * a second look. That keeps a small quota spread across the whole collection.
 */
export async function* harvest(
  windows: ReadonlyArray<readonly [number, number]>,
  perWindow: number,
): AsyncGenerator<RawObject> {
  const total = Math.max(1, perWindow * windows.length);
  const combos: Array<[string, string]> = [];
  for (const century of centuries(windows)) for (const kind of TYPES) combos.push([century, kind]);
  if (combos.length === 0) return;

  const listings = new Map<string, string[]>();
  let yielded = 0;
  const perPass = 2;
  let offset = 0;

  while (yielded < total && offset < 40) {
    let progressed = false;
    for (const [century, kind] of combos) {
      if (yielded >= total) return;
      const key = `${century}|${kind}`;
      if (!listings.has(key)) {
        const url = `${SEARCH}?type=${kind}&imageAvailable=True&creationDate=${century}`;
        try {
          const payload = (await fetchJson(url, { museum: MUSEUM })) as Node | null;
          listings.set(
            key,
            dicts(payload ?? {}, "orderedItems").map((item) => item["id"]).filter(Boolean) as string[],
          );
        } catch (error) {
          if (error instanceof BlockedError) throw error;
          listings.set(key, []);
          continue;
        }
      }
      for (const objectId of (listings.get(key) ?? []).slice(offset, offset + perPass)) {
        progressed = true;
        let record: RawObject | null;
        try {
          record = await loadObject(objectId);
        } catch (error) {
          if (error instanceof BlockedError) throw error;
          continue;
        }
        if (!record) continue;
        yielded += 1;
        yield record;
        if (yielded >= total) return;
      }
    }
    if (!progressed) return;
    offset += perPass;
  }
}
